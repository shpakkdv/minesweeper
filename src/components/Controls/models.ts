import { AppStatus, GameLevel } from 'constant';
import * as ControlsModels from 'containers/Controls/models';

export interface ControlsProps {
  appStatus: AppStatus;
  cellSize: number;
  gameLevel: GameLevel;
  minesFound: number;
  renderWhileSolving: boolean;
  finishGame: ControlsModels.ActionCreator.FinishGame;
  setCellSize: ControlsModels.ActionCreator.SetCellSize;
  setRenderWhileSolvingValue: ControlsModels.ActionCreator.SetRenderWhileSolvingValue;
  solveUntilWin: ControlsModels.ActionCreator.SolveUntilWin;
  startOver: ControlsModels.ActionCreator.StartOver;
  tryToSolveAutomatically: ControlsModels.ActionCreator.TryToSolveAutomatically;
}
