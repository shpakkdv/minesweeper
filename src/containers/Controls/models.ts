import { AppStatus, GameLevel } from 'constant';
import { IAction, IEmptyAction } from 'models';

export interface State {
  gameLevel: GameLevel;
  renderWhileSolving: boolean;
  appStatus: AppStatus;
  // from 5 till 100 px
  cellSize: number;
}

export namespace Payload {
  export interface SetAppStatus {
    appStatus: AppStatus;
  }
  export interface SetCellSize {
    cellSize: number;
  }
  export interface SetGameLevel {
    gameLevel: GameLevel;
  }
  export interface SetRenderWhileSolvingValue {
    renderWhileSolving: boolean;
  }
}

export namespace Action {
  export type FinishGame = IEmptyAction;
  export type SetAppStatus = IAction<Payload.SetAppStatus>;
  export type SetCellSize = IAction<Payload.SetCellSize>;
  export type SetGameLevel = IAction<Payload.SetGameLevel>;
  export type SetRenderWhileSolvingValue = IAction<Payload.SetRenderWhileSolvingValue>;
  export type SolveUntilWin = IEmptyAction;
  export type StartOver = IEmptyAction;
  export type TryToSolveAutomatically = IEmptyAction;
}

export namespace ActionCreator {
  export type FinishGame = () => Action.FinishGame;
  export type SetAppStatus = (appStatus: AppStatus) => Action.SetAppStatus;
  export type SetCellSize = (cellSize: number) => Action.SetCellSize;
  export type SetGameLevel = (gameLevel: GameLevel) => Action.SetGameLevel;
  export type SetRenderWhileSolvingValue = (renderWhileSolving: boolean) => Action.SetRenderWhileSolvingValue;
  export type SolveUntilWin = () => Action.SolveUntilWin;
  export type StartOver = () => Action.StartOver;
  export type TryToSolveAutomatically = () => Action.TryToSolveAutomatically;
}
