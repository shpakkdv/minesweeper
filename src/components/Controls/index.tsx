import clamp from 'lodash/clamp';
import classNames from 'classnames';
import React, { useCallback, useState, useEffect } from 'react';

import { AppStatus, LevelLabelByLevel, MinesNumberByLevel } from 'constant';
import { ControlsProps } from './models';
import styles from './styles.module.scss';

const Controls: React.FunctionComponent<ControlsProps> = ({
  appStatus,
  cellSize,
  gameLevel,
  renderWhileSolving,
  minesFound,
  setCellSize,
  setRenderWhileSolvingValue,
  finishGame,
  solveUntilWin,
  startOver,
  tryToSolveAutomatically,
}) => {
  const [cellSizeValue, setCellSizeValue] = useState(cellSize);
  useEffect(() => {
    setCellSizeValue(cellSize);
  }, [cellSize, setCellSizeValue]);

  const onRenderChanged = useCallback<React.ChangeEventHandler<HTMLInputElement>>(
    () => {
      setRenderWhileSolvingValue(!renderWhileSolving);
    },
    [renderWhileSolving, setRenderWhileSolvingValue],
  );

  const onCellSizeChanged = useCallback<React.ChangeEventHandler<HTMLInputElement>>(
    (event) => {
      const value = Number(event.target.value);

      if (isFinite(value)) {
        setCellSizeValue(value);
      }
    },
    [setCellSizeValue],
  );

  const onCellSizeBlur = useCallback<React.FocusEventHandler<HTMLInputElement>>(() => {
    const newValue = clamp(Number(cellSizeValue), 5, 100);

    if (newValue !== cellSize) {
      setCellSize(newValue);
    }

    setCellSizeValue(newValue);
  }, [cellSize, setCellSize, cellSizeValue, setCellSizeValue]);

  const onCellSizePressEnter = useCallback<React.KeyboardEventHandler<HTMLInputElement>>((event) => {
    if (event.key === 'Enter') {
      onCellSizeBlur(event as any);
    }
  }, [onCellSizeBlur]);

  const notStarted = !gameLevel;
  const canNotSolve = notStarted || [AppStatus.SOLVING, AppStatus.FINISHED].includes(appStatus);

  return (
    <div className={styles.Controls}>
      <div className={styles.controlsContainer}>
        <div className={styles.column}>
          <button onClick={tryToSolveAutomatically} disabled={canNotSolve}>Try to solve automatically</button>
          <button onClick={solveUntilWin} disabled={canNotSolve}>Solve until the win</button>
        </div>
        <div className={styles.column}>
          <button onClick={startOver} disabled={notStarted}>Start over</button>
          <button onClick={finishGame} disabled={notStarted}>Finish the game</button>
        </div>
        <div className={styles.column}>
          <div className={styles.checkbox}>
            <input type="checkbox" id="render" disabled={canNotSolve} onChange={onRenderChanged} checked={renderWhileSolving} />
            <label htmlFor="render">Render while solving</label>
          </div>
          <div className={styles.numericInput}>
            <input
              type="number"
              id="cellSize"
              min="5"
              max="100"
              disabled={notStarted}
              value={cellSizeValue}
              onChange={onCellSizeChanged}
              onBlur={onCellSizeBlur}
              onKeyDown={onCellSizePressEnter}
            />
            <label htmlFor="cellSize">Cell size (5-100px)</label>
          </div>
        </div>
      </div>
      {gameLevel &&
        <div className={classNames(styles.column, styles.info)}>
          <span>{'Level: '}<b>{LevelLabelByLevel[gameLevel]}</b></span>
          <span>{'Mines found: '}<b>{`${minesFound} / ${MinesNumberByLevel[gameLevel]}`}</b></span>
        </div>
      }
    </div>
  );
};

export default Controls;
