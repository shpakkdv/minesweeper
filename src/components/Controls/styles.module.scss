.Controls {
  display: flex;
  justify-content: space-between;
  background-color: antiquewhite;

  & button {
    margin-bottom: 5px;
    cursor: pointer;

    &:disabled {
      cursor: not-allowed;
    }

    &:first-child {
      margin-top: 5px;
    }
  }
}

.controlsContainer {
  display: flex;
}

.column {
  display: flex;
  flex-direction: column;
  width: 200px;
  margin-left: 20px;
}

.checkbox {
  display: flex;
  align-items: center;
  height: 50%;

  & input:disabled {
    cursor: not-allowed;
  }

  & label {
    padding-left: 5px;
    user-select: none;
  }
}

.info {
  margin-right: 10px;
}

.numericInput {
  & input {
    width: 50px;

    &:disabled {
      cursor: not-allowed;
    }
  }

  & label {
    padding-left: 5px;
  }
}
