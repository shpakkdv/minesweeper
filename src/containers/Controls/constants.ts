import { keyMirror } from 'utils/keyMirror';

export const ActionType = keyMirror(
  {
    FINISH_GAME: null,
    SET_GAME_LEVEL: null,
    SET_GAME_LOADING: null,
    SET_RENDER_WHILE_SOLVING_VALUE: null,
    SOLVE_UNTIL_WIN: null,
    START_OVER: null,
    TRY_TO_SOLVE_AUTOMATICALLY: null,
  },
  'Controls',
);
