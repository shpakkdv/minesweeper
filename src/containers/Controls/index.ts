import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import Controls from 'components/Controls';
import { minesFound } from 'containers/GameField/selectors';
import {
  setCellSize,
  tryToSolveAutomatically,
  solveUntilWin,
  startOver,
  finishGame,
  setRenderWhileSolvingValue,
} from './actions';
import { appStatus, cellSize, gameLevel, renderWhileSolving } from './selectors';

const mapStateToProps = createStructuredSelector({
  appStatus,
  cellSize,
  gameLevel,
  renderWhileSolving,
  minesFound,
});

const mapDispatchToProps = {
  finishGame,
  setCellSize,
  setRenderWhileSolvingValue,
  solveUntilWin,
  startOver,
  tryToSolveAutomatically,
};

export default connect(mapStateToProps, mapDispatchToProps)(Controls);
