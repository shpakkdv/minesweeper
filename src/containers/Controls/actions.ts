import { ActionType } from './constants';
import { ActionCreator } from './models';

export const finishGame: ActionCreator.FinishGame = () => ({
  type: ActionType.FINISH_GAME,
});

export const setGameLevel: ActionCreator.SetGameLevel = (gameLevel) => ({
  type: ActionType.SET_GAME_LEVEL,
  payload: {
    gameLevel,
  },
});

export const setAppStatus: ActionCreator.SetAppStatus = (appStatus) => ({
  type: ActionType.SET_APP_STATUS,
  payload: {
    appStatus,
  },
});

export const setCellSize: ActionCreator.SetCellSize = (cellSize) => ({
  type: ActionType.SET_CELL_SIZE,
  payload: {
    cellSize,
  },
});

export const setRenderWhileSolvingValue: ActionCreator.SetRenderWhileSolvingValue = (renderWhileSolving) => ({
  type: ActionType.SET_RENDER_WHILE_SOLVING_VALUE,
  payload: {
    renderWhileSolving,
  },
});

export const solveUntilWin: ActionCreator.SolveUntilWin = () => ({
  type: ActionType.SOLVE_UNTIL_WIN,
});

export const startOver: ActionCreator.StartOver = () => ({
  type: ActionType.START_OVER,
});

export const tryToSolveAutomatically: ActionCreator.TryToSolveAutomatically = () => ({
  type: ActionType.TRY_TO_SOLVE_AUTOMATICALLY,
});
