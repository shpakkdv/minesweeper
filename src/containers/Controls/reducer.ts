import { Reducer } from 'redux';

import { IAction } from 'models';
import { ActionType } from './constants';
import { State, Action } from './models';

const initialState: State = {
  gameLevel: null,
  renderWhileSolving: false,
  appStatus: null,
  cellSize: 40,
};

const reducer: Reducer<State, IAction> = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_GAME_LEVEL:
      return setGameLevel(state, action);
    case ActionType.SET_APP_STATUS:
      return setAppStatus(state, action);
    case ActionType.SET_CELL_SIZE:
      return setCellSize(state, action);
    case ActionType.SET_RENDER_WHILE_SOLVING_VALUE:
      return setRenderWhileSolvingValue(state, action);

    default:
      return state;
  }
};

export default reducer;

export const setGameLevel: Reducer<State, Action.SetGameLevel> = (state, { payload: { gameLevel } }) => ({
  ...state,
  gameLevel,
});

export const setRenderWhileSolvingValue: Reducer<State, Action.SetRenderWhileSolvingValue> = (state, { payload: { renderWhileSolving } }) => ({
  ...state,
  renderWhileSolving,
});

export const setAppStatus: Reducer<State, Action.SetAppStatus> = (state, { payload: { appStatus } }) => ({
  ...state,
  appStatus,
});

export const setCellSize: Reducer<State, Action.SetCellSize> = (state, { payload: { cellSize } }) => ({
  ...state,
  cellSize,
});
