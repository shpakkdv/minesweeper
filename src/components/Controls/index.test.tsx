import React from 'react'
import { shallow } from 'enzyme'

import Controls from './index';

describe('Controls', () => {
  it('Render controls', () => {
    const wrapper = shallow(
      <Controls
        appStatus={null}
        cellSize={40}
        gameLevel={null}
        renderWhileSolving={false}
        minesFound={1}
        setCellSize={jest.fn()}
        setRenderWhileSolvingValue={jest.fn()}
        finishGame={jest.fn()}
        solveUntilWin={jest.fn()}
        startOver={jest.fn()}
        tryToSolveAutomatically={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('All controls should be disabled before game start', () => {
    const wrapper = shallow(
      <Controls
        appStatus={null}
        cellSize={40}
        gameLevel={null}
        renderWhileSolving={false}
        minesFound={1}
        setCellSize={jest.fn()}
        setRenderWhileSolvingValue={jest.fn()}
        finishGame={jest.fn()}
        solveUntilWin={jest.fn()}
        startOver={jest.fn()}
        tryToSolveAutomatically={jest.fn()}
      />
    );

    const buttons = wrapper.find('button');
    const inputs = wrapper.find('input');

    expect(buttons.length).toBe(4);
    expect(buttons.everyWhere(button => button.props().disabled)).toBe(true);
    expect(inputs.length).toBe(2);
    expect(inputs.everyWhere(input => input.props().disabled)).toBe(true);
  });
});
