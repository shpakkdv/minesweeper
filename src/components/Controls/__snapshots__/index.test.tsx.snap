// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Controls Render controls 1`] = `
<div
  className="Controls"
>
  <div
    className="controlsContainer"
  >
    <div
      className="column"
    >
      <button
        disabled={true}
        onClick={[MockFunction]}
      >
        Try to solve automatically
      </button>
      <button
        disabled={true}
        onClick={[MockFunction]}
      >
        Solve until the win
      </button>
    </div>
    <div
      className="column"
    >
      <button
        disabled={true}
        onClick={[MockFunction]}
      >
        Start over
      </button>
      <button
        disabled={true}
        onClick={[MockFunction]}
      >
        Finish the game
      </button>
    </div>
    <div
      className="column"
    >
      <div
        className="checkbox"
      >
        <input
          checked={false}
          disabled={true}
          id="render"
          onChange={[Function]}
          type="checkbox"
        />
        <label
          htmlFor="render"
        >
          Render while solving
        </label>
      </div>
      <div
        className="numericInput"
      >
        <input
          disabled={true}
          id="cellSize"
          max="100"
          min="5"
          onBlur={[Function]}
          onChange={[Function]}
          onKeyDown={[Function]}
          type="number"
          value={40}
        />
        <label
          htmlFor="cellSize"
        >
          Cell size (5-100px)
        </label>
      </div>
    </div>
  </div>
</div>
`;
